src/validation_functions.jl

```diff
@@ -0,0 +1,144 @@
+struct ValidationInfo
+    field_descriptor::Dict
+    struct_name::AbstractString
+    ps_struct::PowerSystemType
+    field_type::Any
+    limits::NamedTuple{(:min, :max), Tuple{Union{Nothing, Float64}, Union{Nothing, Float64}}}
+end
+
+#Get validation info for one struct.
+function get_config_descriptor(config::Vector, name::AbstractString)
+    for item in config
+        if item["struct_name"] == name
+            return item
+        end
+    end
+    error("PowerSystems struct $name does not exist in validation configuration file")
+end
+
+#Get validation info for one field of one struct.
+function get_field_descriptor(struct_descriptor::Dict, fieldname::AbstractString)
+    for field in struct_descriptor["fields"]
+        if field["name"] == fieldname
+            return field
+        end
+    end
+    error("field $fieldname does not exist in $(struct_descriptor["name"])")
+end
+
+function validate_fields(sys::System, ps_struct::T) where T <: PowerSystemType
+    struct_descriptor = get_config_descriptor(sys.validation_descriptor, repr(T))
+    is_valid = true
+
+    for (name,fieldtype) in zip(fieldnames(T), fieldtypes(T))
+        field_value = getfield(ps_struct, name)
+        if isnothing(field_value) #Many structs are of type Union{Nothing, xxx}.
+            ;
+        elseif fieldtype <: Union{Nothing, PowerSystemType} && !(fieldtype <: Component)
+            # Recurse. Components are validated separately and do not need to be validated twice.
+            if !validate_fields(sys, getfield(ps_struct, name))
+                is_valid = false
+            end
+        else
```

Comment:
If I'm reading this correctly, if fieldtype is a Component then it will fall into this else block.  But then "valid_range" won't be defined and it will be skipped.  Is that what you want?

Reply:
yes